// appium-mac-driver/build/lib/appium-for-mac.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumBaseDriver = require('appium-base-driver');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _teen_process = require('teen_process');

var _appiumSupport = require('appium-support');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var DEFAULT_A4M_HOST = '127.0.0.1';
var DEFAULT_A4M_PORT = 4622;

var REQ_A4M_APP_PATH = _path2['default'].resolve('/Applications', 'AppiumForMac.app');

var a4mLog = _appiumSupport.logger.getLogger('Appium4Mac');

var AppiumForMac = (function () {
  function AppiumForMac() {
    _classCallCheck(this, AppiumForMac);

    this.proxyHost = DEFAULT_A4M_HOST;
    this.proxyPort = DEFAULT_A4M_PORT;
    this.proc = null;
    this.jwproxy = new _appiumBaseDriver.JWProxy({ server: this.proxyHost, port: this.proxyPort });
  }

  _createClass(AppiumForMac, [{
    key: 'start',
    value: function start() {
      var startDetector, processIsAlive, a4mBinary, _arr, _loop, _i;

      return _regeneratorRuntime.async(function start$(context$2$0) {
        var _this = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(REQ_A4M_APP_PATH));

          case 2:
            if (context$2$0.sent) {
              context$2$0.next = 4;
              break;
            }

            throw new Error("Could not verify AppiumForMacDriver install; please install to your /Applications folder");

          case 4:
            startDetector = function startDetector(stdout, stderr) {
              return stderr.indexOf("Started HTTP server") !== -1;
            };

            processIsAlive = false;
            context$2$0.prev = 6;
            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(this.killAll());

          case 9:
            a4mBinary = _path2['default'].resolve(REQ_A4M_APP_PATH, "Contents", "MacOS", "AppiumForMac");

            this.proc = new _teen_process.SubProcess(a4mBinary, []);
            processIsAlive = true;

            // handle log output
            _arr = ['STDOUT', 'STDERR'];

            _loop = function () {
              var stream = _arr[_i];
              _this.proc.on('lines-' + stream.toLowerCase(), function (lines) {
                var _iteratorNormalCompletion = true;
                var _didIteratorError = false;
                var _iteratorError = undefined;

                try {
                  for (var _iterator = _getIterator(lines), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var l = _step.value;

                    a4mLog.info('[' + stream + '] ' + l.trim());
                  }
                } catch (err) {
                  _didIteratorError = true;
                  _iteratorError = err;
                } finally {
                  try {
                    if (!_iteratorNormalCompletion && _iterator['return']) {
                      _iterator['return']();
                    }
                  } finally {
                    if (_didIteratorError) {
                      throw _iteratorError;
                    }
                  }
                }
              });
            };

            for (_i = 0; _i < _arr.length; _i++) {
              _loop();
            }

            // handle out-of-bound exit by simply logging
            // TODO add ability for driver to handle this gracefully and maybe
            // restart
            this.proc.on('exit', function (code, signal) {
              processIsAlive = false;
              var msg = 'AppiumForMac exited unexpectedly with code ' + code + ', ' + ('signal ' + signal);
              _logger2['default'].error(msg);
            });
            _logger2['default'].info('Spawning AppiumForMac with: ' + this.appium4macdriver);

            // start subproc and wait for startDetector
            context$2$0.next = 19;
            return _regeneratorRuntime.awrap(this.proc.start(startDetector));

          case 19:
            context$2$0.next = 21;
            return _regeneratorRuntime.awrap(this.waitForOnline());

          case 21:
            context$2$0.next = 30;
            break;

          case 23:
            context$2$0.prev = 23;
            context$2$0.t0 = context$2$0['catch'](6);

            this.emit(AppiumForMac.EVENT_ERROR, context$2$0.t0);
            // just because we had an error doesn't mean the winappdriver process
            // finished; we should clean up if necessary

            if (!processIsAlive) {
              context$2$0.next = 29;
              break;
            }

            context$2$0.next = 29;
            return _regeneratorRuntime.awrap(this.proc.stop());

          case 29:
            _logger2['default'].errorAndThrow(context$2$0.t0);

          case 30:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[6, 23]]);
    }
  }, {
    key: 'sessionId',
    value: function sessionId() {
      if (this.state !== AppiumForMac.STATE_ONLINE) {
        return null;
      }

      return this.jwproxy.sessionId;
    }
  }, {
    key: 'waitForOnline',
    value: function waitForOnline() {
      return _regeneratorRuntime.async(function waitForOnline$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', true);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getStatus',
    value: function getStatus() {
      return _regeneratorRuntime.async(function getStatus$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.sendCommand('/status', 'GET'));

          case 2:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'startSession',
    value: function startSession(caps) {
      return _regeneratorRuntime.async(function startSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.sendCommand('/session', 'POST', { desiredCapabilities: caps }));

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'stop',
    value: function stop() {
      return _regeneratorRuntime.async(function stop$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.prev = 0;

            if (!this.proc) {
              context$2$0.next = 4;
              break;
            }

            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.proc.stop());

          case 4:
            context$2$0.next = 9;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](0);

            _logger2['default'].error(context$2$0.t0);

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[0, 6]]);
    }
  }, {
    key: 'sendCommand',
    value: function sendCommand(url, method, body) {
      var res;
      return _regeneratorRuntime.async(function sendCommand$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            res = undefined;
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.jwproxy.command(url, method, body));

          case 4:
            res = context$2$0.sent;
            context$2$0.next = 11;
            break;

          case 7:
            context$2$0.prev = 7;
            context$2$0.t0 = context$2$0['catch'](1);

            if (!(context$2$0.t0.message.indexOf("Did not get a valid response object") === -1 || context$2$0.t0.message.indexOf("value") !== -1)) {
              context$2$0.next = 11;
              break;
            }

            throw context$2$0.t0;

          case 11:
            return context$2$0.abrupt('return', res);

          case 12:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 7]]);
    }
  }, {
    key: 'proxyReq',
    value: function proxyReq(req, res) {
      return _regeneratorRuntime.async(function proxyReq$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.jwproxy.proxyReqRes(req, res));

          case 2:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'killAll',
    value: function killAll() {
      var processName;
      return _regeneratorRuntime.async(function killAll$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            processName = "AppiumForMac";

            // js hint cannot handle backticks, even escaped, within template literals
            _logger2['default'].info('Killing any old AppiumForMac');
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(_appiumSupport.process.killProcess(processName));

          case 4:
            _logger2['default'].info("Successfully cleaned up old Appium4Mac servers");

          case 5:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'deleteSession',
    value: function deleteSession() {
      return _regeneratorRuntime.async(function deleteSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug('Deleting AppiumForMac server session');
            // rely on jwproxy's intelligence to know what we're talking about and
            // delete the current session
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.sendCommand('/', 'DELETE'));

          case 4:
            context$2$0.next = 9;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](1);

            _logger2['default'].warn('Did not get confirmation AppiumForMac deleteSession worked; ' + ('Error was: ' + context$2$0.t0));

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 6]]);
    }
  }]);

  return AppiumForMac;
})();

exports.AppiumForMac = AppiumForMac;
exports.DEFAULT_A4M_HOST = DEFAULT_A4M_HOST;
exports.DEFAULT_A4M_PORT = DEFAULT_A4M_PORT;
exports['default'] = AppiumForMac;

// set up our subprocess object

// TODO: Actually check via HTTP

// need to cover over A4M's bad handling of responses, which sometimes
// don't have 'value' properties
