// appium-instruments/build/lib/streams.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _through = require('through');

var _through2 = _interopRequireDefault(_through);

function clearBufferChars(output) {
  // Instruments output is buffered, so for each log output we also output
  // a stream of very many ****. This function strips those out so all we
  // get is the log output we care about
  var re = /(\n|^)\*+\n?/g;
  output = output.toString();
  output = output.replace(re, '');
  return output;
}

function outputStream() {
  return (0, _through2['default'])(function outputStreamHandler(output) {
    output = clearBufferChars(output);

    // if we have multiple log lines, indent non-first ones
    if (output !== '') {
      // get rid of newline at the end
      output = output.replace(/\n$/m, '');
      // indent instead of newline
      output = output.replace(/\n/m, '\n       ');
      output = '[INST] ' + output;
      output = output.green;
      _logger2['default'].debug(output);
    }
    this.queue(output);
  });
}

function errorStream() {
  return (0, _through2['default'])(function errorStreamHandler(output) {
    // get rid of newline at the end
    output = output.replace(/\n$/m, '');
    output = '[INST STDERR] ' + output;
    output = output.yellow;
    _logger2['default'].debug(output);
    this.queue(output);
  });
}

function webSocketAlertStream(webSocket) {
  return (0, _through2['default'])(function webSocketAlertStreamHandler(output) {
    if (webSocket) {
      var re = /Call to onAlert returned 'YES'/;
      if (re.test(output)) {
        _logger2['default'].debug('Emiting alert message...');
        webSocket.sockets.emit('alert', { message: output });
      }
    }
    this.queue(output);
  });
}

function dumpStream() {
  return (0, _through2['default'])(function nullStreamHandler() {/* ignore all stream output */});
}

exports.outputStream = outputStream;
exports.errorStream = errorStream;
exports.webSocketAlertStream = webSocketAlertStream;
exports.dumpStream = dumpStream;
