// appium-base-driver/build/test/jsonwp-proxy/mock-request.js
'use strict';

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});
function resFixture(url, method, json) {
  if (/\/status$/.test(url)) {
    return [200, { status: 0, value: { foo: 'bar' } }];
  }
  if (/\/element\/bad\/text$/.test(url)) {
    return [500, { status: 11, value: { message: 'Invisible element' } }];
  }
  if (/\/element\/200\/text$/.test(url)) {
    return [200, { status: 11, value: { message: 'Invisible element' } }];
  }
  if (/\/element\/200\/value$/.test(url)) {
    return [200, { status: 0, sessionId: 'innersessionid', value: 'foobar' }];
  }
  if (/\/session$/.test(url) && method === "POST") {
    if (json.desiredCapabilities && json.desiredCapabilities.redirect) {
      return [303, 'http://localhost:4444/wd/hub/session/123'];
    } else {
      return [200, { status: 0, sessionId: '123', value: { browserName: 'boo' } }];
    }
  }
  if (/\/nochrome$/.test(url)) {
    return [100, { status: 0, value: { message: 'chrome not reachable' } }];
  }
  throw new Error("Can't handle url " + url);
}

function request(opts) {
  var _resFixture, _resFixture2, statusCode, body, response;

  return _regeneratorRuntime.async(function request$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!/badurl$/.test(opts.url)) {
          context$1$0.next = 2;
          break;
        }

        throw new Error("noworky");

      case 2:
        _resFixture = resFixture(opts.url, opts.method, opts.json);
        _resFixture2 = _slicedToArray(_resFixture, 2);
        statusCode = _resFixture2[0];
        body = _resFixture2[1];
        response = {
          statusCode: statusCode,
          headers: { 'Content-type': 'application/json' },
          body: body
        };
        return context$1$0.abrupt('return', response);

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports['default'] = request;
module.exports = exports['default'];
