// appium-fake-driver/build/lib/fake-element.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _xmldom = require('xmldom');

var _xmldom2 = _interopRequireDefault(_xmldom);

var FakeElement = (function () {
  function FakeElement(xmlNode, app) {
    _classCallCheck(this, FakeElement);

    this.app = app;
    this.node = xmlNode;
    this.nodeAttrs = {};
    this.type = this.node.tagName;
    this.attrs = {};
    this.css = {};
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _getIterator(_lodash2['default'].values(this.node.attributes)), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var _step$value = _step.value;
        var _name = _step$value.name;
        var value = _step$value.value;

        this.nodeAttrs[_name] = value;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator['return']) {
          _iterator['return']();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    this.parseCss();
  }

  _createClass(FakeElement, [{
    key: 'parseCss',
    value: function parseCss() {
      if (this.nodeAttrs.style) {
        var segments = this.nodeAttrs.style.split(";");
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          for (var _iterator2 = _getIterator(segments), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var s = _step2.value;

            var _s$split = s.split(":");

            var _s$split2 = _slicedToArray(_s$split, 2);

            var prop = _s$split2[0];
            var val = _s$split2[1];

            prop = prop.trim();
            val = val.trim();
            this.css[prop] = val;
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2['return']) {
              _iterator2['return']();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }
      }
    }
  }, {
    key: 'setAttr',
    value: function setAttr(k, v) {
      this.attrs[k] = v;
    }
  }, {
    key: 'getAttr',
    value: function getAttr(k) {
      return this.attrs[k] || "";
    }
  }, {
    key: 'isVisible',
    value: function isVisible() {
      return this.nodeAttrs.visible !== "false";
    }
  }, {
    key: 'isEnabled',
    value: function isEnabled() {
      return this.nodeAttrs.enabled !== "false";
    }
  }, {
    key: 'isSelected',
    value: function isSelected() {
      return this.nodeAttrs.selected === "true";
    }
  }, {
    key: 'getLocation',
    value: function getLocation() {
      return {
        x: parseFloat(this.nodeAttrs.left || 0),
        y: parseFloat(this.nodeAttrs.top || 0)
      };
    }
  }, {
    key: 'getSize',
    value: function getSize() {
      return {
        width: parseFloat(this.nodeAttrs.width || 0),
        height: parseFloat(this.nodeAttrs.height || 0)
      };
    }
  }, {
    key: 'click',
    value: function click() {
      var curClicks = this.getAttr('clicks') || 0;
      this.setAttr('clicks', curClicks + 1);
      var alertId = this.nodeAttrs.showAlert;
      if (alertId) {
        this.app.showAlert(alertId);
      }
    }
  }, {
    key: 'equals',
    value: function equals(other) {
      return this.node === other.node;
    }
  }, {
    key: 'hasPrompt',
    value: function hasPrompt() {
      return this.nodeAttrs.hasPrompt === "true";
    }
  }, {
    key: 'getCss',
    value: function getCss(prop) {
      if (_lodash2['default'].has(this.css, prop)) {
        return this.css[prop];
      }
      return null;
    }
  }, {
    key: 'tagName',
    get: function get() {
      return this.node.tagName;
    }
  }, {
    key: 'xmlFragment',
    get: function get() {
      var frag = new _xmldom2['default'].XMLSerializer().serializeToString(this.node);
      return new _xmldom2['default'].DOMParser().parseFromString(frag, "application/xml");
    }
  }]);

  return FakeElement;
})();

exports.FakeElement = FakeElement;
