// appium-support/build/lib/plist.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _plist = require('plist');

var _plist2 = _interopRequireDefault(_plist);

var _bplistCreator = require('bplist-creator');

var _bplistCreator2 = _interopRequireDefault(_bplistCreator);

var _bplistParser = require('bplist-parser');

var _bplistParser2 = _interopRequireDefault(_bplistParser);

var _fs = require('./fs');

var _fs2 = _interopRequireDefault(_fs);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var parseFile = _bluebird2['default'].promisify(_bplistParser2['default'].parseFile);

// XML Plist library helper
function parseXmlPlistFile(plistFilename) {
  var xmlContent;
  return _regeneratorRuntime.async(function parseXmlPlistFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_fs2['default'].readFile(plistFilename, 'utf8'));

      case 2:
        xmlContent = context$1$0.sent;
        return context$1$0.abrupt('return', _plist2['default'].parse(xmlContent));

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function parsePlistFile(plist) {
  var mustExist = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];
  var quiet = arguments.length <= 2 || arguments[2] === undefined ? true : arguments[2];
  var obj, type;
  return _regeneratorRuntime.async(function parsePlistFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_fs2['default'].exists(plist));

      case 2:
        if (context$1$0.sent) {
          context$1$0.next = 9;
          break;
        }

        if (!mustExist) {
          context$1$0.next = 7;
          break;
        }

        _logger2['default'].errorAndThrow('Plist file doesn\'t exist: \'' + plist + '\'');
        context$1$0.next = 9;
        break;

      case 7:
        _logger2['default'].debug('Plist file \'' + plist + '\' does not exist. Returning an empty plist.');
        return context$1$0.abrupt('return', {});

      case 9:
        obj = {};
        type = 'binary';
        context$1$0.prev = 11;
        context$1$0.next = 14;
        return _regeneratorRuntime.awrap(parseFile(plist));

      case 14:
        obj = context$1$0.sent;

        if (!obj.length) {
          context$1$0.next = 19;
          break;
        }

        obj = obj[0];
        context$1$0.next = 20;
        break;

      case 19:
        throw new Error('Binary file \'' + plist + '\'\' appears to be empty');

      case 20:
        context$1$0.next = 34;
        break;

      case 22:
        context$1$0.prev = 22;
        context$1$0.t0 = context$1$0['catch'](11);
        context$1$0.prev = 24;
        context$1$0.next = 27;
        return _regeneratorRuntime.awrap(parseXmlPlistFile(plist));

      case 27:
        obj = context$1$0.sent;

        type = 'xml';
        context$1$0.next = 34;
        break;

      case 31:
        context$1$0.prev = 31;
        context$1$0.t1 = context$1$0['catch'](24);

        _logger2['default'].errorAndThrow('Could not parse plist file \'' + plist + '\' as XML: ' + context$1$0.t1.message);

      case 34:

        if (!quiet) {
          _logger2['default'].debug('Parsed plist file \'' + plist + '\' as ' + type);
        }
        return context$1$0.abrupt('return', obj);

      case 36:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[11, 22], [24, 31]]);
}

function updatePlistFile(plist, updatedFields) {
  var binary = arguments.length <= 2 || arguments[2] === undefined ? true : arguments[2];
  var mustExist = arguments.length <= 3 || arguments[3] === undefined ? true : arguments[3];
  var quiet = arguments.length <= 4 || arguments[4] === undefined ? true : arguments[4];
  var obj, newPlist;
  return _regeneratorRuntime.async(function updatePlistFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        obj = undefined;
        context$1$0.prev = 1;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(parsePlistFile(plist, mustExist));

      case 4:
        obj = context$1$0.sent;
        context$1$0.next = 10;
        break;

      case 7:
        context$1$0.prev = 7;
        context$1$0.t0 = context$1$0['catch'](1);

        _logger2['default'].errorAndThrow('Could not update plist: ' + context$1$0.t0.message);

      case 10:
        _lodash2['default'].extend(obj, updatedFields);
        newPlist = binary ? (0, _bplistCreator2['default'])(obj) : _plist2['default'].build(obj);
        context$1$0.prev = 12;
        context$1$0.next = 15;
        return _regeneratorRuntime.awrap(_fs2['default'].writeFile(plist, newPlist));

      case 15:
        context$1$0.next = 20;
        break;

      case 17:
        context$1$0.prev = 17;
        context$1$0.t1 = context$1$0['catch'](12);

        _logger2['default'].errorAndThrow('Could not save plist: ' + context$1$0.t1.message);

      case 20:
        if (!quiet) {
          _logger2['default'].debug('Wrote plist file \'' + plist + '\'');
        }

      case 21:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[1, 7], [12, 17]]);
}

exports.parsePlistFile = parsePlistFile;
exports.updatePlistFile = updatePlistFile;

// handle nonexistant file
