// appium-uiauto/build/lib/build-script.js
'use strict';

var _defineProperty = require('babel-runtime/helpers/define-property')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _appiumSupport = require('appium-support');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var rootDir = _path2['default'].resolve(__dirname, '..', 'uiauto');
if (!__dirname.match(/build\/lib/)) {
  rootDir = _path2['default'].resolve(__dirname, 'uiauto');
}

// this regex helps us get the file path of an import
var importRe = /^#import ('|")([^('|")]+)('|")$/mg;

function getDepsForFile(file, filesExamined) {
  var extraImports = arguments.length <= 2 || arguments[2] === undefined ? [] : arguments[2];

  var data, deps, imports, match, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, importedFile, importedPath, importedDeps;

  return _regeneratorRuntime.async(function getDepsForFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!_lodash2['default'].includes(filesExamined, file)) {
          context$1$0.next = 2;
          break;
        }

        throw new Error('Re-examining file ' + file + '; you need to make sure ' + 'the graph is set up so we do not require files twice');

      case 2:

        // save the file so we don't look at it again
        filesExamined.push(file);
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(file, 'utf8'));

      case 5:
        data = context$1$0.sent;
        deps = _defineProperty({}, file, []);
        imports = [];
        match = importRe.exec(data);

        while (match) {
          if (match) {
            imports.push(match[2]);
          }
          match = importRe.exec(data);
        }
        // add in any extra imports sent in
        imports = extraImports.concat(imports);

        // go through all the imports for the file, and do the same process
        _iteratorNormalCompletion = true;
        _didIteratorError = false;
        _iteratorError = undefined;
        context$1$0.prev = 14;
        _iterator = _getIterator(imports);

      case 16:
        if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
          context$1$0.next = 26;
          break;
        }

        importedFile = _step.value;
        importedPath = _path2['default'].resolve(_path2['default'].dirname(file), importedFile);
        context$1$0.next = 21;
        return _regeneratorRuntime.awrap(getDepsForFile(importedPath, filesExamined));

      case 21:
        importedDeps = context$1$0.sent;

        deps[file].push(importedDeps);

      case 23:
        _iteratorNormalCompletion = true;
        context$1$0.next = 16;
        break;

      case 26:
        context$1$0.next = 32;
        break;

      case 28:
        context$1$0.prev = 28;
        context$1$0.t0 = context$1$0['catch'](14);
        _didIteratorError = true;
        _iteratorError = context$1$0.t0;

      case 32:
        context$1$0.prev = 32;
        context$1$0.prev = 33;

        if (!_iteratorNormalCompletion && _iterator['return']) {
          _iterator['return']();
        }

      case 35:
        context$1$0.prev = 35;

        if (!_didIteratorError) {
          context$1$0.next = 38;
          break;
        }

        throw _iteratorError;

      case 38:
        return context$1$0.finish(35);

      case 39:
        return context$1$0.finish(32);

      case 40:
        return context$1$0.abrupt('return', deps);

      case 41:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[14, 28, 32, 40], [33,, 35, 39]]);
}

function buildScriptFromDeps(deps) {
  var script, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _step2$value, file, subDepsArray, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, subDeps, fileContents, newFileData, fileWithoutRoot;

  return _regeneratorRuntime.async(function buildScriptFromDeps$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        script = '';
        _iteratorNormalCompletion2 = true;
        _didIteratorError2 = false;
        _iteratorError2 = undefined;
        context$1$0.prev = 4;
        _iterator2 = _getIterator(_lodash2['default'].toPairs(deps));

      case 6:
        if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
          context$1$0.next = 48;
          break;
        }

        _step2$value = _slicedToArray(_step2.value, 2);
        file = _step2$value[0];
        subDepsArray = _step2$value[1];
        _iteratorNormalCompletion3 = true;
        _didIteratorError3 = false;
        _iteratorError3 = undefined;
        context$1$0.prev = 13;
        _iterator3 = _getIterator(subDepsArray);

      case 15:
        if (_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done) {
          context$1$0.next = 23;
          break;
        }

        subDeps = _step3.value;
        context$1$0.next = 19;
        return _regeneratorRuntime.awrap(buildScriptFromDeps(subDeps));

      case 19:
        script += context$1$0.sent;

      case 20:
        _iteratorNormalCompletion3 = true;
        context$1$0.next = 15;
        break;

      case 23:
        context$1$0.next = 29;
        break;

      case 25:
        context$1$0.prev = 25;
        context$1$0.t0 = context$1$0['catch'](13);
        _didIteratorError3 = true;
        _iteratorError3 = context$1$0.t0;

      case 29:
        context$1$0.prev = 29;
        context$1$0.prev = 30;

        if (!_iteratorNormalCompletion3 && _iterator3['return']) {
          _iterator3['return']();
        }

      case 32:
        context$1$0.prev = 32;

        if (!_didIteratorError3) {
          context$1$0.next = 35;
          break;
        }

        throw _iteratorError3;

      case 35:
        return context$1$0.finish(32);

      case 36:
        return context$1$0.finish(29);

      case 37:
        context$1$0.next = 39;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(file, 'utf8'));

      case 39:
        fileContents = context$1$0.sent;
        newFileData = stripImports(fileContents);
        fileWithoutRoot = file.replace(rootDir + '/', '');

        script += '\n/* begin file: ' + fileWithoutRoot + ' */\n';
        script += newFileData;
        script += '\n/* end file: ' + fileWithoutRoot + ' */\n';

      case 45:
        _iteratorNormalCompletion2 = true;
        context$1$0.next = 6;
        break;

      case 48:
        context$1$0.next = 54;
        break;

      case 50:
        context$1$0.prev = 50;
        context$1$0.t1 = context$1$0['catch'](4);
        _didIteratorError2 = true;
        _iteratorError2 = context$1$0.t1;

      case 54:
        context$1$0.prev = 54;
        context$1$0.prev = 55;

        if (!_iteratorNormalCompletion2 && _iterator2['return']) {
          _iterator2['return']();
        }

      case 57:
        context$1$0.prev = 57;

        if (!_didIteratorError2) {
          context$1$0.next = 60;
          break;
        }

        throw _iteratorError2;

      case 60:
        return context$1$0.finish(57);

      case 61:
        return context$1$0.finish(54);

      case 62:
        return context$1$0.abrupt('return', script);

      case 63:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[4, 50, 54, 62], [13, 25, 29, 37], [30,, 32, 36], [55,, 57, 61]]);
}

function stripImports(data) {
  // get rid of the import statements from the string
  data = data.replace(importRe, '');
  data = data.trim();
  return data;
}

function buildScript(entryPoint) {
  var extraImports = arguments.length <= 1 || arguments[1] === undefined ? [] : arguments[1];
  var filesExamined, deps, script;
  return _regeneratorRuntime.async(function buildScript$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        filesExamined = [];
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(getDepsForFile(entryPoint, filesExamined, extraImports));

      case 3:
        deps = context$1$0.sent;
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(buildScriptFromDeps(deps));

      case 6:
        script = context$1$0.sent;
        return context$1$0.abrupt('return', script);

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports['default'] = buildScript;
module.exports = exports['default'];

// make sure we don't have a cycle in our dependencies

// check for import statements in the file, and make a list of them

// recursively get dependencies for imported files

// go through all the dependencies and recursively
// add them to our script string

// keep a list of files we examine for the purposes of making sure we're
// not adding the same file to the collated version twice
