// appium-ios-driver/build/lib/app-utils.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumSupport = require('appium-support');

var _loggerJs = require('./logger.js');

var _loggerJs2 = _interopRequireDefault(_loggerJs);

var plistObjects = {};

function extractPlist(app) {
  var plistFile;
  return _regeneratorRuntime.async(function extractPlist$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        plistFile = _path2['default'].resolve(app, 'Info.plist');
        context$1$0.prev = 1;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(_appiumSupport.plist.parsePlistFile(plistFile));

      case 4:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 7:
        context$1$0.prev = 7;
        context$1$0.t0 = context$1$0['catch'](1);

        _loggerJs2['default'].errorAndThrow('Could not extract Info.plist from application: ' + context$1$0.t0.message);

      case 10:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[1, 7]]);
}

function extractPlistEntry(app, entryName) {
  return _regeneratorRuntime.async(function extractPlistEntry$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!_lodash2['default'].isUndefined(plistObjects[app])) {
          context$1$0.next = 4;
          break;
        }

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(extractPlist(app));

      case 3:
        plistObjects[app] = context$1$0.sent;

      case 4:
        return context$1$0.abrupt('return', plistObjects[app][entryName]);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function extractAppDisplayName(app) {
  return _regeneratorRuntime.async(function extractAppDisplayName$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _loggerJs2['default'].debug("Getting application name from app");
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(extractPlistEntry(app, 'CFBundleDisplayName'));

      case 3:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function extractBundleId(app) {
  var bundleId;
  return _regeneratorRuntime.async(function extractBundleId$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(extractPlistEntry(app, 'CFBundleIdentifier'));

      case 2:
        bundleId = context$1$0.sent;

        _loggerJs2['default'].debug('Getting bundle ID from app \'' + app + '\': \'' + bundleId + '\'');
        return context$1$0.abrupt('return', bundleId);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.extractAppDisplayName = extractAppDisplayName;
exports.extractBundleId = extractBundleId;
