// appium-ios-simulator/build/lib/touch-enroll.js
'use strict';

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var touchEnrollMenuKeys = ['Toggle Enrolled State', 'Touch ID Enrolled'];
var touchEnrollBackups = undefined;
var NS_USER_KEY_EQUIVALENTS = 'NSUserKeyEquivalents';
var TOUCH_ENROLL_KEY_CODE = '@~$^t';

function setTouchEnrollKey() {
  var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, key;

  return _regeneratorRuntime.async(function setTouchEnrollKey$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(backupTouchEnrollShortcuts());

      case 2:
        _iteratorNormalCompletion = true;
        _didIteratorError = false;
        _iteratorError = undefined;
        context$1$0.prev = 5;
        _iterator = _getIterator(touchEnrollMenuKeys);

      case 7:
        if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
          context$1$0.next = 14;
          break;
        }

        key = _step.value;
        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(setUserDefault(NS_USER_KEY_EQUIVALENTS, key, TOUCH_ENROLL_KEY_CODE));

      case 11:
        _iteratorNormalCompletion = true;
        context$1$0.next = 7;
        break;

      case 14:
        context$1$0.next = 20;
        break;

      case 16:
        context$1$0.prev = 16;
        context$1$0.t0 = context$1$0['catch'](5);
        _didIteratorError = true;
        _iteratorError = context$1$0.t0;

      case 20:
        context$1$0.prev = 20;
        context$1$0.prev = 21;

        if (!_iteratorNormalCompletion && _iterator['return']) {
          _iterator['return']();
        }

      case 23:
        context$1$0.prev = 23;

        if (!_didIteratorError) {
          context$1$0.next = 26;
          break;
        }

        throw _iteratorError;

      case 26:
        return context$1$0.finish(23);

      case 27:
        return context$1$0.finish(20);

      case 28:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[5, 16, 20, 28], [21,, 23, 27]]);
}

function setTouchEnrollKeys(pairs) {
  var _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _step2$value, key, value;

  return _regeneratorRuntime.async(function setTouchEnrollKeys$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _iteratorNormalCompletion2 = true;
        _didIteratorError2 = false;
        _iteratorError2 = undefined;
        context$1$0.prev = 3;
        _iterator2 = _getIterator(pairs);

      case 5:
        if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
          context$1$0.next = 14;
          break;
        }

        _step2$value = _slicedToArray(_step2.value, 2);
        key = _step2$value[0];
        value = _step2$value[1];
        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(setUserDefault(NS_USER_KEY_EQUIVALENTS, key, value));

      case 11:
        _iteratorNormalCompletion2 = true;
        context$1$0.next = 5;
        break;

      case 14:
        context$1$0.next = 20;
        break;

      case 16:
        context$1$0.prev = 16;
        context$1$0.t0 = context$1$0['catch'](3);
        _didIteratorError2 = true;
        _iteratorError2 = context$1$0.t0;

      case 20:
        context$1$0.prev = 20;
        context$1$0.prev = 21;

        if (!_iteratorNormalCompletion2 && _iterator2['return']) {
          _iterator2['return']();
        }

      case 23:
        context$1$0.prev = 23;

        if (!_didIteratorError2) {
          context$1$0.next = 26;
          break;
        }

        throw _iteratorError2;

      case 26:
        return context$1$0.finish(23);

      case 27:
        return context$1$0.finish(20);

      case 28:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[3, 16, 20, 28], [21,, 23, 27]]);
}

function getTouchEnrollKeys() {
  var backups, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, key;

  return _regeneratorRuntime.async(function getTouchEnrollKeys$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        backups = [];
        _iteratorNormalCompletion3 = true;
        _didIteratorError3 = false;
        _iteratorError3 = undefined;
        context$1$0.prev = 4;
        _iterator3 = _getIterator(touchEnrollMenuKeys);

      case 6:
        if (_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done) {
          context$1$0.next = 18;
          break;
        }

        key = _step3.value;
        context$1$0.t0 = backups;
        context$1$0.t1 = key;
        context$1$0.next = 12;
        return _regeneratorRuntime.awrap(getUserDefault(NS_USER_KEY_EQUIVALENTS, key));

      case 12:
        context$1$0.t2 = context$1$0.sent;
        context$1$0.t3 = [context$1$0.t1, context$1$0.t2];
        context$1$0.t0.push.call(context$1$0.t0, context$1$0.t3);

      case 15:
        _iteratorNormalCompletion3 = true;
        context$1$0.next = 6;
        break;

      case 18:
        context$1$0.next = 24;
        break;

      case 20:
        context$1$0.prev = 20;
        context$1$0.t4 = context$1$0['catch'](4);
        _didIteratorError3 = true;
        _iteratorError3 = context$1$0.t4;

      case 24:
        context$1$0.prev = 24;
        context$1$0.prev = 25;

        if (!_iteratorNormalCompletion3 && _iterator3['return']) {
          _iterator3['return']();
        }

      case 27:
        context$1$0.prev = 27;

        if (!_didIteratorError3) {
          context$1$0.next = 30;
          break;
        }

        throw _iteratorError3;

      case 30:
        return context$1$0.finish(27);

      case 31:
        return context$1$0.finish(24);

      case 32:
        return context$1$0.abrupt('return', backups);

      case 33:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[4, 20, 24, 32], [25,, 27, 31]]);
}

/**
 * Get MacOS User Defaults by domain and key (for reference: `man defaults`)
 * @param {*} domain {string}
 * @param {*} key {string|number|boolean}
 */
function getUserDefault(domain, key) {
  var res, stdout, nsUserKeyArr, _iteratorNormalCompletion4, _didIteratorError4, _iteratorError4, _iterator4, _step4, _step4$value, testKey, value;

  return _regeneratorRuntime.async(function getUserDefault$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        res = undefined;
        context$1$0.prev = 1;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('defaults', ['read', 'Apple Global Domain', domain]));

      case 4:
        res = context$1$0.sent;
        context$1$0.next = 10;
        break;

      case 7:
        context$1$0.prev = 7;
        context$1$0.t0 = context$1$0['catch'](1);
        return context$1$0.abrupt('return');

      case 10:
        stdout = res.stdout;
        nsUserKeyArr = stdout.trim().replace(/^{/, '') // Remove leading {
        .replace(/}$/, '') // Remove trailing }
        .trim().replace(/;$/, '') // Remove trailing semicolon
        .split(';') // Break up expressions by semicolon
        .map(function (expr) {
          var _expr$split = expr.split('=');

          var _expr$split2 = _slicedToArray(_expr$split, 2);

          var key = _expr$split2[0];
          var value = _expr$split2[1];

          key = key.trim().replace(/^"/, '').replace(/"$/, '');
          value = value.trim().replace(/^"/, '').replace(/"$/, '');
          return [key, value];
        });
        _iteratorNormalCompletion4 = true;
        _didIteratorError4 = false;
        _iteratorError4 = undefined;
        context$1$0.prev = 15;
        _iterator4 = _getIterator(nsUserKeyArr);

      case 17:
        if (_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done) {
          context$1$0.next = 26;
          break;
        }

        _step4$value = _slicedToArray(_step4.value, 2);
        testKey = _step4$value[0];
        value = _step4$value[1];

        if (!(testKey === key)) {
          context$1$0.next = 23;
          break;
        }

        return context$1$0.abrupt('return', value.replace(/\\\\/g, '\\'));

      case 23:
        _iteratorNormalCompletion4 = true;
        context$1$0.next = 17;
        break;

      case 26:
        context$1$0.next = 32;
        break;

      case 28:
        context$1$0.prev = 28;
        context$1$0.t1 = context$1$0['catch'](15);
        _didIteratorError4 = true;
        _iteratorError4 = context$1$0.t1;

      case 32:
        context$1$0.prev = 32;
        context$1$0.prev = 33;

        if (!_iteratorNormalCompletion4 && _iterator4['return']) {
          _iterator4['return']();
        }

      case 35:
        context$1$0.prev = 35;

        if (!_didIteratorError4) {
          context$1$0.next = 38;
          break;
        }

        throw _iteratorError4;

      case 38:
        return context$1$0.finish(35);

      case 39:
        return context$1$0.finish(32);

      case 40:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[1, 7], [15, 28, 32, 40], [33,, 35, 39]]);
}

/**
 * Sets a MacOS User Default value on a domain
 * @param {*} domain 
 * @param {*} key 
 * @param {*} value 
 */
function setUserDefault(domain, key, value) {
  return _regeneratorRuntime.async(function setUserDefault$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('defaults', ['write', 'Apple Global Domain', domain, '-dict-add', key, typeof value === 'undefined' ? 'nil' : value]));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function backupTouchEnrollShortcuts() {
  return _regeneratorRuntime.async(function backupTouchEnrollShortcuts$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (touchEnrollBackups) {
          context$1$0.next = 4;
          break;
        }

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(getTouchEnrollKeys());

      case 3:
        touchEnrollBackups = context$1$0.sent;

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function restoreTouchEnrollShortcuts() {
  return _regeneratorRuntime.async(function restoreTouchEnrollShortcuts$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!touchEnrollBackups) {
          context$1$0.next = 4;
          break;
        }

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(setTouchEnrollKeys(touchEnrollBackups));

      case 3:
        touchEnrollBackups = undefined;

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function getTouchEnrollBackups() {
  return touchEnrollBackups;
}

exports.setTouchEnrollKey = setTouchEnrollKey;
exports.getTouchEnrollKeys = getTouchEnrollKeys;
exports.setTouchEnrollKeys = setTouchEnrollKeys;
exports.getUserDefault = getUserDefault;
exports.setUserDefault = setUserDefault;
exports.getTouchEnrollBackups = getTouchEnrollBackups;
exports.touchEnrollMenuKeys = touchEnrollMenuKeys;
exports.backupTouchEnrollShortcuts = backupTouchEnrollShortcuts;
exports.restoreTouchEnrollShortcuts = restoreTouchEnrollShortcuts;
exports.NS_USER_KEY_EQUIVALENTS = NS_USER_KEY_EQUIVALENTS;
exports.TOUCH_ENROLL_KEY_CODE = TOUCH_ENROLL_KEY_CODE;

// If it doesn't find anything for this domain, it will throw an error so return undefined

// Parse the result into a Javascript array
