// appium-android-driver/build/lib/commands/lua/element.js
"use strict";

var _regeneratorRuntime = require("babel-runtime/regenerator")["default"];

var _Object$assign = require("babel-runtime/core-js/object/assign")["default"];

Object.defineProperty(exports, "__esModule", {
    value: true
});

var commands = {},
    helpers = {};

/*
 * 点击lua元素 2017/4/24
 *
 */
commands.luaClick = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.testbundle.sendAction("element:click", { elementId: elementId }));

            case 2:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 3:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 * 获取元素属性 (文本: text)
 *
 */
commands.getLuaAttribute = function callee$0$0(attribute, elementId) {
    var p;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                p = { attribute: attribute, elementId: elementId };
                context$1$0.next = 3;
                return _regeneratorRuntime.awrap(this.testbundle.sendAction("element:getAttribute", p));

            case 3:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 4:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 * 设置元素的值 (文本)
 *
 */
commands.setLuaValue = function callee$0$0(keys, elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.setLuaElementValue(keys, elementId, false));

            case 2:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 3:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 *  元素是否displaed
 *
 */
commands.luaElementDisplayed = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.getLuaAttribute("displayed", elementId));

            case 2:
                context$1$0.t0 = context$1$0.sent;
                return context$1$0.abrupt("return", context$1$0.t0 === 'true');

            case 4:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 * 元素是否enabled
 *
 */
commands.luaElementEnabled = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.getLuaAttribute("enabled", elementId));

            case 2:
                context$1$0.t0 = context$1$0.sent;
                return context$1$0.abrupt("return", context$1$0.t0 === 'true');

            case 4:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 * 元素是否selected
 *
 */
commands.luaElementSelected = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.getLuaAttribute("selected", elementId));

            case 2:
                context$1$0.t0 = context$1$0.sent;
                return context$1$0.abrupt("return", context$1$0.t0 === 'true');

            case 4:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 * 元素 location( x, y)
 *
 */
commands.getLuaLocation = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.testbundle.sendAction("element:getLocation", { elementId: elementId }));

            case 2:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 3:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

commands.getLuaLocationInView = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.getLuaLocation(elementId));

            case 2:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 3:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

/*
 * 元素 size( width, height)
 */
commands.getLuaSize = function callee$0$0(elementId) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.testbundle.sendAction("element:getSize", { elementId: elementId }));

            case 2:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 3:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

helpers.setLuaElementValue = function callee$0$0(keys, elementId) {
    var replace = arguments.length <= 2 || arguments[2] === undefined ? false : arguments[2];
    var text, params;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                text = keys;

                if (keys instanceof Array) {
                    text = keys.join("");
                }
                params = {
                    elementId: elementId,
                    text: text,
                    replace: replace,
                    unicodeKeyboard: this.opts.unicodeKeyboard
                };
                context$1$0.next = 5;
                return _regeneratorRuntime.awrap(this.testbundle.sendAction("element:setText", params));

            case 5:
                return context$1$0.abrupt("return", context$1$0.sent);

            case 6:
            case "end":
                return context$1$0.stop();
        }
    }, null, this);
};

_Object$assign(commands, helpers);
exports["default"] = commands;
module.exports = exports["default"];
