// appium-fake-driver/build/test/helpers.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Promise = require('babel-runtime/core-js/promise')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _wd = require('wd');

var _wd2 = _interopRequireDefault(_wd);

var TEST_HOST = 'localhost';
var TEST_PORT = 4774;
var TEST_APP = _path2['default'].resolve(__dirname, "..", "..", "test", "fixtures", "app.xml");

var DEFAULT_CAPS = { platformName: 'Fake', deviceName: 'Fake', app: TEST_APP };

function initSession(caps) {
  var _this = this;

  var resolve = function resolve() {};
  var driver = undefined;
  before(function callee$1$0() {
    return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          driver = _wd2['default'].promiseChainRemote({ host: TEST_HOST, port: TEST_PORT });
          context$2$0.next = 3;
          return _regeneratorRuntime.awrap(driver.init(caps));

        case 3:
          resolve(driver);

        case 4:
        case 'end':
          return context$2$0.stop();
      }
    }, null, _this);
  });
  after(function callee$1$0() {
    return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          context$2$0.next = 2;
          return _regeneratorRuntime.awrap(driver.quit());

        case 2:
        case 'end':
          return context$2$0.stop();
      }
    }, null, _this);
  });
  return new _Promise(function (_resolve) {
    resolve = _resolve;
  });
}

exports.initSession = initSession;
exports.TEST_APP = TEST_APP;
exports.TEST_HOST = TEST_HOST;
exports.TEST_PORT = TEST_PORT;
exports.DEFAULT_CAPS = DEFAULT_CAPS;
