// appium-android-driver/build/lib/commands/coverage.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _temp = require('temp');

var _temp2 = _interopRequireDefault(_temp);

var _appiumSupport = require('appium-support');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var commands = {},
    helpers = {},
    extensions = {};

function unlinkFile(file) {
  return _regeneratorRuntime.async(function unlinkFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(file));

      case 2:
        if (!context$1$0.sent) {
          context$1$0.next = 5;
          break;
        }

        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.unlink(file));

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

commands.endCoverage = function callee$0$0(intentToBroadcast, ecOnDevicePath) {
  var localFile, b64data, data;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        localFile = _temp2['default'].path({ prefix: 'appium', suffix: '.ec' });
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(unlinkFile(localFile));

      case 3:
        b64data = '';
        context$1$0.prev = 4;
        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(this.adb.rimraf(ecOnDevicePath));

      case 7:
        context$1$0.next = 9;
        return _regeneratorRuntime.awrap(this.adb.broadcastProcessEnd(intentToBroadcast, this.appProcess));

      case 9:
        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(this.adb.pull(ecOnDevicePath, localFile));

      case 11:
        context$1$0.next = 13;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(localFile));

      case 13:
        data = context$1$0.sent;

        b64data = new Buffer(data).toString('base64');
        context$1$0.next = 17;
        return _regeneratorRuntime.awrap(unlinkFile(localFile));

      case 17:
        context$1$0.next = 22;
        break;

      case 19:
        context$1$0.prev = 19;
        context$1$0.t0 = context$1$0['catch'](4);

        _logger2['default'].debug('Error ending test coverage: ' + context$1$0.t0.message);

      case 22:
        return context$1$0.abrupt('return', b64data);

      case 23:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[4, 19]]);
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports['default'] = extensions;

// ensure the ec we're pulling is newly created as a result of the intent.
