// appium-ios-driver/build/test/e2e/helpers/session.js
"use strict";

var _createClass = require("babel-runtime/helpers/create-class")["default"];

var _classCallCheck = require("babel-runtime/helpers/class-call-check")["default"];

var _getIterator = require("babel-runtime/core-js/get-iterator")["default"];

var _regeneratorRuntime = require("babel-runtime/regenerator")["default"];

var _interopRequireDefault = require("babel-runtime/helpers/interop-require-default")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _env = require('./env');

var _env2 = _interopRequireDefault(_env);

var _bluebird = require("bluebird");

var _bluebird2 = _interopRequireDefault(_bluebird);

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _2 = require('../../..');

var _appiumBaseDriver = require('appium-base-driver');

var _libLogger = require('../../../lib/logger');

var _libLogger2 = _interopRequireDefault(_libLogger);

var MOCHA_TIMEOUT = 60 * 1000 * (process.env.TRAVIS ? 20 : 4);
var MOCHA_SAFARI_TIMEOUT = MOCHA_TIMEOUT * 2;

var Session = (function () {
  function Session() {
    var _this = this;

    var desired = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var opts = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, Session);

    this.desired = desired;
    this.opts = opts;
    this.initialized = false;

    this.rawDriver = new _2.IosDriver(this.opts);

    // wrapping the driver so that the call goes
    // through executeCommand
    this.driver = {};
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      var _loop = function () {
        var c = _step.value;

        _this.driver[c] = (function () {
          var _rawDriver;

          for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
          }

          return (_rawDriver = this.rawDriver).executeCommand.apply(_rawDriver, [c].concat(args));
        }).bind(_this);
      };

      for (var _iterator = _getIterator(_appiumBaseDriver.ALL_COMMANDS), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        _loop();
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"]) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    var _arr = ['createSession', 'deleteSession'];
    for (var _i = 0; _i < _arr.length; _i++) {
      var c = _arr[_i];
      this.driver[c] = this.rawDriver[c].bind(this.rawDriver);
    }
  }

  _createClass(Session, [{
    key: "setUp",
    value: function setUp() {
      var caps, init, attempts;
      return _regeneratorRuntime.async(function setUp$(context$2$0) {
        var _this2 = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            caps = _lodash2["default"].clone(this.desired);

            _lodash2["default"].defaults(caps, _env2["default"].CAPS);

            _libLogger2["default"].debug("caps -->", caps);
            _libLogger2["default"].debug("opts -->", this.opts);

            init = function init(remainingAttempts) {
              return _regeneratorRuntime.async(function init$(context$3$0) {
                while (1) switch (context$3$0.prev = context$3$0.next) {
                  case 0:
                    _libLogger2["default"].debug("remainingAttempts --> " + remainingAttempts);
                    context$3$0.prev = 1;
                    context$3$0.next = 4;
                    return _regeneratorRuntime.awrap(this.driver.createSession(caps));

                  case 4:
                    context$3$0.next = 20;
                    break;

                  case 6:
                    context$3$0.prev = 6;
                    context$3$0.t0 = context$3$0["catch"](1);

                    _libLogger2["default"].debug("Init failed with error --> " + context$3$0.t0);
                    remainingAttempts--;

                    if (!(remainingAttempts === 0)) {
                      context$3$0.next = 14;
                      break;
                    }

                    throw context$3$0.t0;

                  case 14:
                    context$3$0.next = 16;
                    return _regeneratorRuntime.awrap(this.driver.deleteSession());

                  case 16:
                    context$3$0.next = 18;
                    return _regeneratorRuntime.awrap(_bluebird2["default"].delay(10000));

                  case 18:
                    context$3$0.next = 20;
                    return _regeneratorRuntime.awrap(init(remainingAttempts));

                  case 20:
                  case "end":
                    return context$3$0.stop();
                }
              }, null, _this2, [[1, 6]]);
            };

            attempts = this.opts['no-retry'] ? 1 : 3;

            if (_env2["default"].MAX_RETRY) attempts = Math.min(_env2["default"].MAX_RETRY, attempts);
            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(init(attempts));

          case 9:
            this.initialized = true;
            context$2$0.next = 12;
            return _regeneratorRuntime.awrap(this.driver.implicitWait(_env2["default"].IMPLICIT_WAIT_TIMEOUT));

          case 12:
          case "end":
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: "tearDown",
    value: function tearDown() {
      return _regeneratorRuntime.async(function tearDown$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!this.initialized) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.driver.deleteSession());

          case 4:
            context$2$0.next = 9;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0["catch"](1);

            _libLogger2["default"].warn("didn't quit cleanly.");

          case 9:
          case "end":
            return context$2$0.stop();
        }
      }, null, this, [[1, 6]]);
    }
  }]);

  return Session;
})();

exports.Session = Session;
exports.MOCHA_TIMEOUT = MOCHA_TIMEOUT;
exports.MOCHA_SAFARI_TIMEOUT = MOCHA_SAFARI_TIMEOUT;
