// appium-ios-driver/build/test/e2e/helpers/webview.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _env = require('./env');

var _env2 = _interopRequireDefault(_env);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _nodeUuid = require('node-uuid');

var _nodeUuid2 = _interopRequireDefault(_nodeUuid);

var BROWSERS = ['safari'];

function spinTitle(expTitle, browser) {
  var timeout = arguments.length <= 2 || arguments[2] === undefined ? 90 : arguments[2];

  var _curTitle = arguments.length <= 3 || arguments[3] === undefined ? undefined : arguments[3];

  var pageTitle;
  return _regeneratorRuntime.async(function spinTitle$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(timeout <= 0)) {
          context$1$0.next = 2;
          break;
        }

        throw new Error('Title never became \'' + expTitle + '\'. Last known title was \'' + _curTitle + '\'');

      case 2:
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(browser.title());

      case 4:
        pageTitle = context$1$0.sent;

        if (!(pageTitle.indexOf(expTitle) < 0)) {
          context$1$0.next = 11;
          break;
        }

        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(_bluebird2['default'].delay(500));

      case 8:
        context$1$0.next = 10;
        return _regeneratorRuntime.awrap(spinTitle(expTitle, browser, timeout - 1, pageTitle));

      case 10:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 11:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function loadWebView(desired, browser, urlToLoad, titleToSpin) {
  var app, uuid, ctxs, url;
  return _regeneratorRuntime.async(function loadWebView$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        app = typeof desired === 'object' ? desired.app || desired.browserName : desired;
        uuid = _nodeUuid2['default'].v1();

        if (typeof urlToLoad === 'undefined') {
          urlToLoad = _env2['default'].GUINEA_TEST_END_POINT + '?' + uuid;
        }
        if (typeof titleToSpin === 'undefined') {
          titleToSpin = uuid;
        }

        if (!(BROWSERS.indexOf(app) > -1)) {
          context$1$0.next = 12;
          break;
        }

        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(browser.setUrl(urlToLoad));

      case 7:
        context$1$0.next = 9;
        return _regeneratorRuntime.awrap(_bluebird2['default'].delay(3000));

      case 9:
        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(spinTitle(titleToSpin, browser));

      case 11:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 12:
        context$1$0.next = 14;
        return _regeneratorRuntime.awrap(browser.contexts());

      case 14:
        ctxs = context$1$0.sent;

        ctxs.length.should.be.above(0);
        context$1$0.next = 18;
        return _regeneratorRuntime.awrap(browser.context(ctxs[1]));

      case 18:
        context$1$0.next = 20;
        return _regeneratorRuntime.awrap(browser.url());

      case 20:
        url = context$1$0.sent;

        if (!(url !== urlToLoad)) {
          context$1$0.next = 24;
          break;
        }

        context$1$0.next = 24;
        return _regeneratorRuntime.awrap(browser.setUrl(urlToLoad));

      case 24:
        context$1$0.next = 26;
        return _regeneratorRuntime.awrap(spinTitle(titleToSpin, browser));

      case 26:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 27:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function spinWait(spinFn) {
  var waitMs = arguments.length <= 1 || arguments[1] === undefined ? 10000 : arguments[1];
  var intMs = arguments.length <= 2 || arguments[2] === undefined ? 500 : arguments[2];
  var begunAt, endAt, spin;
  return _regeneratorRuntime.async(function spinWait$(context$1$0) {
    var _this2 = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        begunAt = Date.now();
        endAt = begunAt + waitMs;

        spin = function spin() {
          return _regeneratorRuntime.async(function spin$(context$2$0) {
            var _this = this;

            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                context$2$0.prev = 0;
                context$2$0.next = 3;
                return _regeneratorRuntime.awrap(spinFn());

              case 3:
                context$2$0.next = 10;
                break;

              case 5:
                context$2$0.prev = 5;
                context$2$0.t0 = context$2$0['catch'](0);

                if (!(Date.now() > endAt)) {
                  context$2$0.next = 9;
                  break;
                }

                throw new Error('spinWait condition unfulfilled. Promise rejected with error: ' + context$2$0.t0);

              case 9:
                return context$2$0.abrupt('return', setTimeout(function callee$3$0() {
                  return _regeneratorRuntime.async(function callee$3$0$(context$4$0) {
                    while (1) switch (context$4$0.prev = context$4$0.next) {
                      case 0:
                        context$4$0.next = 2;
                        return _regeneratorRuntime.awrap(spin(spinFn));

                      case 2:
                        return context$4$0.abrupt('return', context$4$0.sent);

                      case 3:
                      case 'end':
                        return context$4$0.stop();
                    }
                  }, null, _this);
                }, intMs));

              case 10:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this2, [[0, 5]]);
        };

        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(spin(spinFn));

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function skip(reason, done) {
  console.warn('skipping: ' + reason); //eslint-disable-line no-console
  return done();
}

exports.spinTitle = spinTitle;
exports.loadWebView = loadWebView;
exports.skip = skip;
exports.spinWait = spinWait;
